ARG still not working properly in FROM

### Description
This was already discussed extensively since 2017 at #859 and partially fixed by #1299 with a new release today. I tested it, and there are several cases where it still does not work.

MWE pom.xml:
```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>com.example</groupId>
    <artifactId>dmp-test</artifactId>
    <version>1.0.0-SNAPSHOT</version>
    <packaging>pom</packaging>

    <properties>
        <java.version>1.8</java.version>
        <resource.delimiter>@</resource.delimiter>
        <maven.compiler.source>${java.version}</maven.compiler.source>
        <maven.compiler.target>${java.version}</maven.compiler.target>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
        <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>

        <busybox.version>latest</busybox.version>
    </properties>

    <build>
        <plugins>
            <plugin>
                <groupId>io.fabric8</groupId>
                <artifactId>docker-maven-plugin</artifactId>
                <version>0.34.0</version>
                <configuration>
                    <images>
                        <image>
                            <alias>${project.artifactId}</alias>
                            <name>${project.artifactId}:${project.version}</name>
                            <build>
                                <contextDir>${project.basedir}</contextDir>
                                <args>
                                    <busyboxVersion>${busybox.version}</busyboxVersion>
                                    <busyboxVersionName>${busybox.version}</busyboxVersionName>
                                </args>
                            </build>
                        </image>
                    </images>
                </configuration>
            </plugin>
        </plugins>
    </build>
</project>
```

I want to use busyboxVersion, as supplied by Maven, in my Dockerfile, so I do this:
```
ARG busyboxVersion
FROM busybox:${busyboxVersion}
```

But upon building, I get this:
```
[INFO] DOCKER> [dmp-test:1.0.0-SNAPSHOT] "dmp-test": Created docker-build.tar in 83 milliseconds
[ERROR] DOCKER> Unable to check image [busybox:${busyboxVersion}] : {"message":"no such image: busybox:${busyboxVersion}: invalid reference format"} (Bad Request: 400) [{"message":"no such image: busybox:${busyboxVersion}: invalid reference format"} (Bad Request: 400)]
```

Apparently, the ARG is not resolved, although this is the correct format as per [the documentation](https://docs.docker.com/engine/reference/builder/#understand-how-arg-and-from-interact). So, I change it to this:
```
ARG busyboxVersion
FROM busybox:$busyboxVersion
```

And get this:
```
[INFO] DOCKER> [dmp-test:1.0.0-SNAPSHOT] "dmp-test": Created docker-build.tar in 122 milliseconds
[ERROR] DOCKER> Unable to check image [busybox:] : {"message":"no such image: busybox:: invalid reference format"} (Bad Request: 400) [{"message":"no such image: busybox:: invalid reference format"} (Bad Request: 400)]
```

The ARG is now resolved, but has no content. Hence, I test this:
```
ARG busyboxVersion
FROM busybox:latest
ARG busyboxVersionName
LABEL busyboxVersion=${busyboxVersion}
LABEL busyboxVersionName=${busyboxVersionName}
```

The image is now built (since the version is again hard-coded), but upon inspecting it in Docker, I get this:
```
"Labels": {
    "busyboxVersion": "",
    "busyboxVersionName": "latest"
}
```

Once again, the ARG was resolved to an empty string, while the second ARG was interpreted correctly. I further this test like so:
```
FROM busybox:latest
ARG busyboxVersion
ARG busyboxVersionName
LABEL busyboxVersion=${busyboxVersion}
LABEL busyboxVersionName=${busyboxVersionName}
```

Now inspecting the Docker image, everything is alright:
```
"Labels": {
    "busyboxVersion": "latest",
    "busyboxVersionName": "latest"
}
```

### Info

* d-m-p version : 0.34.0
* Maven version (`mvn -v`) :
```
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: C:\Program Files\apache-maven-3.6.3\bin\..
Java version: 11.0.7, vendor: AdoptOpenJDK, runtime: C:\Program Files\AdoptOpenJDK\jdk-11.0.7.10-hotspot
Default locale: en_US, platform encoding: Cp1252
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"
```
* Docker version : 19.03.12
* If it's a bug, how to reproduce : see above
* If it's a feature request, what is your use case : -
* Sample project : see above

Thanks for taking this on, I will also note here that the variable format with the braces does not work irrespective of the POM usage, such as in this:
```
ARG busyboxVersion=latest
FROM busybox:${busyboxVersion}
```
```
[INFO] DOCKER> [dmp-test:1.0.0-SNAPSHOT] "dmp-test": Created docker-build.tar in 160 milliseconds
[ERROR] DOCKER> Unable to check image [busybox:${busyboxVersion}] : {"message":"no such image: busybox:${busyboxVersion}: invalid reference format"} (Bad Request: 400) [{"message":"no such image: busybox:${busyboxVersion}: invalid reference format"} (Bad Request: 400)]
```

--- a/doc/changelog.md
+++ b/doc/changelog.md
@@ -1,5 +1,8 @@
 # ChangeLog
 
+* 0.34-SNAPSHOT
+  - Fix #1372: Plugin now resolves ARG provided in BuildImageConfiguration
+
 * **0.34.0** (2020-09-13)
   - Support `ARG` in `FROM` ([#859](https://github.com/fabric8io/docker-maven-plugin/issues/859))
   - Handle authentication tokens returned from credential helpers ([#1348](https://github.com/fabric8io/docker-maven-plugin/issues/1348))

--- a/src/main/java/io/fabric8/maven/docker/service/BuildService.java
+++ b/src/main/java/io/fabric8/maven/docker/service/BuildService.java
@@ -406,7 +406,8 @@ public class BuildService {
             File fullDockerFilePath = buildConfig.getAbsoluteDockerFilePath(buildContext.getMojoParameters());
             fromImage = DockerFileUtil.extractBaseImages(
                 fullDockerFilePath,
-                DockerFileUtil.createInterpolator(buildContext.getMojoParameters(), buildConfig.getFilter()));
+                DockerFileUtil.createInterpolator(buildContext.getMojoParameters(), buildConfig.getFilter()),
+                buildConfig.getArgs());
         } catch (IOException e) {
             // Cant extract base image, so we wont try an auto pull. An error will occur later anyway when
             // building the image, so we are passive here.

--- a/src/main/java/io/fabric8/maven/docker/util/DockerFileUtil.java
+++ b/src/main/java/io/fabric8/maven/docker/util/DockerFileUtil.java
@@ -53,9 +53,9 @@ public class DockerFileUtil {
      * @param interpolator interpolator for replacing properties
      * @return LinkedList of base images name or empty collection if none is found.
      */
-    public static List<String> extractBaseImages(File dockerFile, FixedStringSearchInterpolator interpolator) throws IOException {
+    public static List<String> extractBaseImages(File dockerFile, FixedStringSearchInterpolator interpolator, Map<String, String> argsFromBuildConfig) throws IOException {
         List<String[]> fromLines = extractLines(dockerFile, "FROM", interpolator);
-        Map<String, String> args = extractArgs(dockerFile, interpolator);
+        Map<String, String> args = extractArgs(dockerFile, argsFromBuildConfig, interpolator);
         Set<String> result = new LinkedHashSet<>();
         Set<String> fromAlias = new HashSet<>();
         for (String[] fromLine :  fromLines) {
@@ -80,8 +80,8 @@ public class DockerFileUtil {
      * @param interpolator interpolator for replacement
      * @return HashMap of arguments or empty collection if none is found
      */
-    public static Map<String, String> extractArgs(File dockerfile, FixedStringSearchInterpolator interpolator) throws IOException {
-        return extractArgsFromLines(extractLines(dockerfile, "ARG", interpolator));
+    public static Map<String, String> extractArgs(File dockerfile, Map<String, String> argsFromBuildConfig, FixedStringSearchInterpolator interpolator) throws IOException {
+        return extractArgsFromLines(extractLines(dockerfile, "ARG", interpolator), argsFromBuildConfig);
     }
 
     /**
@@ -155,14 +155,23 @@ public class DockerFileUtil {
      * @param argLines list of string arrays containing lines with words
      * @return map of parsed arguments
      */
-    protected static Map<String, String> extractArgsFromLines(List<String[]> argLines) {
+    static Map<String, String> extractArgsFromLines(List<String[]> argLines, Map<String, String> argsFromBuildConfig) {
         Map<String, String> result = new HashMap<>();
         for (String[] argLine : argLines) {
             if (argLine.length > 1) {
-                updateMapWithArgValue(result, argLine[1]);
+                updateMapWithArgValue(result, argsFromBuildConfig, argLine[1]);
             }
         }
         return result;
+    }
+
+    static String resolveArgValueFromStrContainingArgKey(String argString, Map<String, String> args) {
+        if (argString.startsWith("$") && args.containsKey(argString.substring(1))) {
+            return args.get(argString.substring(1));
+        } else if (argString.startsWith("${") && argString.endsWith("}") && args.containsKey(argString.substring(2, argString.length() - 1))) {
+            return args.get(argString.substring(2, argString.length() - 1));
+        }
+        return null;
     }
 
     private static String resolveImageTagFromArgs(String imageTagString, Map<String, String> args) {
@@ -181,13 +190,6 @@ public class DockerFileUtil {
             }
         }
         return imageTagString;
-    }
-
-    private static String resolveArgValueFromStrContainingArgKey(String argString, Map<String, String> args) {
-        if (argString.startsWith("$") && args.containsKey(argString.substring(1))) {
-            return args.get(argString.substring(1));
-        }
-        return null;
     }
 
     private static Reader getFileReaderFromDir(File file) {
@@ -248,7 +250,7 @@ public class DockerFileUtil {
         return new File(homeDir);
     }
 
-    private static void updateMapWithArgValue(Map<String, String> result, String argString) {
+    private static void updateMapWithArgValue(Map<String, String> result, Map<String, String> args, String argString) {
         if (argString.contains("=") || argString.contains(":")) {
             String[] argStringParts = argString.split("[=:]");
             String argStringValue = argString.substring(argStringParts[0].length() + 1);
@@ -261,8 +263,16 @@ public class DockerFileUtil {
             result.put(argStringParts[0], argStringValue);
         } else {
             validateArgValue(argString);
-            result.put(argString.split("\\s+")[0], "");
+            result.putAll(fetchArgsFromBuildConfiguration(argString, args));
         }
+    }
+
+    private static Map<String, String> fetchArgsFromBuildConfiguration(String argString, Map<String, String> args) {
+        Map<String, String> argFromBuildConfig = new HashMap<>();
+        if (args != null) {
+            argFromBuildConfig.put(argString, args.getOrDefault(argString, ""));
+        }
+        return argFromBuildConfig;
     }
 
     private static void validateArgValue(String argStringParam) {

--- a/src/test/java/io/fabric8/maven/docker/util/DockerFileUtilTest.java
+++ b/src/test/java/io/fabric8/maven/docker/util/DockerFileUtilTest.java
@@ -43,14 +43,14 @@ public class DockerFileUtilTest {
     public void testSimple() throws Exception {
         File toTest = copyToTempDir("Dockerfile_from_simple");
         assertEquals("fabric8/s2i-java", DockerFileUtil.extractBaseImages(
-            toTest, FixedStringSearchInterpolator.create()).get(0));
+            toTest, FixedStringSearchInterpolator.create(), Collections.emptyMap()).get(0));
     }
 
     @Test
     public void testMultiStage() throws Exception {
         File toTest = copyToTempDir("Dockerfile_multi_stage");
         Iterator<String> fromClauses = DockerFileUtil.extractBaseImages(
-             toTest, FixedStringSearchInterpolator.create()).iterator();
+             toTest, FixedStringSearchInterpolator.create(), Collections.emptyMap()).iterator();
 
         assertEquals("fabric8/s2i-java", fromClauses.next());
         assertEquals("fabric8/s1i-java", fromClauses.next());
@@ -61,7 +61,7 @@ public class DockerFileUtilTest {
     public void testMultiStageNamed() throws Exception {
         File toTest = copyToTempDir("Dockerfile_multi_stage_named_build_stages");
         Iterator<String> fromClauses = DockerFileUtil.extractBaseImages(
-                toTest, FixedStringSearchInterpolator.create()).iterator();
+                toTest, FixedStringSearchInterpolator.create(), Collections.emptyMap()).iterator();
 
         assertEquals("fabric8/s2i-java", fromClauses.next());
         assertFalse(fromClauses.hasNext());
@@ -71,7 +71,7 @@ public class DockerFileUtilTest {
     public void testMultiStageWithArgs() throws Exception {
         File toTest = copyToTempDir("Dockerfile_multi_stage_with_args");
         Iterator<String> fromClauses = DockerFileUtil.extractBaseImages(
-                toTest, FixedStringSearchInterpolator.create()).iterator();
+                toTest, FixedStringSearchInterpolator.create(), Collections.emptyMap()).iterator();
 
         assertEquals("fabric8/s2i-java:latest", fromClauses.next());
         assertEquals("busybox:latest", fromClauses.next());
@@ -80,52 +80,59 @@ public class DockerFileUtilTest {
 
     @Test
     public void testExtractArgsFromDockerfile() {
-        assertEquals("{VERSION=latest, FULL_IMAGE=busybox:latest}", DockerFileUtil.extractArgsFromLines(Arrays.asList(new String[]{"ARG", "VERSION:latest"}, new String[] {"ARG", "FULL_IMAGE=busybox:latest"})).toString());
-        assertEquals("{user1=someuser, buildno=1}", DockerFileUtil.extractArgsFromLines(Arrays.asList(new String[]{"ARG", "user1=someuser"}, new String[]{"ARG", "buildno=1"})).toString());
-        assertEquals("{NPM_VERSION=latest, NODE_VERSION=latest}", DockerFileUtil.extractArgsFromLines(Arrays.asList(new String[]{"ARG","NODE_VERSION=\"latest\""}, new String[]{"ARG",  "NPM_VERSION=\"latest\""})).toString());
-        assertEquals("{NPM_VERSION=latest, NODE_VERSION=latest}", DockerFileUtil.extractArgsFromLines(Arrays.asList(new String[]{"ARG","NODE_VERSION='latest'"}, new String[]{"ARG",  "NPM_VERSION='latest'"})).toString());
-        assertEquals("{MESSAGE=argument with spaces}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[] {"ARG", "MESSAGE='argument with spaces'"})).toString());
-        assertEquals("{MESSAGE=argument with spaces}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[] {"ARG", "MESSAGE=\"argument with spaces\""})).toString());
-        assertEquals("{TARGETPLATFORM=}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "TARGETPLATFORM"})).toString());
-        assertEquals("{TARGETPLATFORM=}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "TARGETPLATFORM="})).toString());
-        assertEquals("{TARGETPLATFORM=}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "TARGETPLATFORM:"})).toString());
-        assertEquals("{MESSAGE=argument:two}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "MESSAGE=argument:two"})).toString());
-        assertEquals("{MESSAGE2=argument=two}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "MESSAGE2=argument=two"})).toString());
-        assertEquals("{VER=0.0.3}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "VER=0.0.3"})).toString());
-        assertEquals("{VER={0.0.3}}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "VER={0.0.3}"})).toString());
-        assertEquals("{VER=[0.0.3]}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "VER=[0.0.3]"})).toString());
-        assertEquals("{VER={5,6}}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "VER={5,6}"})).toString());
-        assertEquals("{VER={5,6}}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "VER={5,6}"})).toString());
-        assertEquals("{VER={}}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "VER={}"})).toString());
-        assertEquals("{VER=====}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "VER====="})).toString());
-        assertEquals("{MESSAGE=:message}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "MESSAGE=:message"})).toString());
-        assertEquals("{MYAPP_IMAGE=myorg/myapp:latest}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "MYAPP_IMAGE=myorg/myapp:latest"})).toString());
+        assertEquals("{VERSION=latest, FULL_IMAGE=busybox:latest}", DockerFileUtil.extractArgsFromLines(Arrays.asList(new String[]{"ARG", "VERSION:latest"}, new String[] {"ARG", "FULL_IMAGE=busybox:latest"}), Collections.emptyMap()).toString());
+        assertEquals("{user1=someuser, buildno=1}", DockerFileUtil.extractArgsFromLines(Arrays.asList(new String[]{"ARG", "user1=someuser"}, new String[]{"ARG", "buildno=1"}), Collections.emptyMap()).toString());
+        assertEquals("{NPM_VERSION=latest, NODE_VERSION=latest}", DockerFileUtil.extractArgsFromLines(Arrays.asList(new String[]{"ARG","NODE_VERSION=\"latest\""}, new String[]{"ARG",  "NPM_VERSION=\"latest\""}), Collections.emptyMap()).toString());
+        assertEquals("{NPM_VERSION=latest, NODE_VERSION=latest}", DockerFileUtil.extractArgsFromLines(Arrays.asList(new String[]{"ARG","NODE_VERSION='latest'"}, new String[]{"ARG",  "NPM_VERSION='latest'"}), Collections.emptyMap()).toString());
+        assertEquals("{MESSAGE=argument with spaces}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[] {"ARG", "MESSAGE='argument with spaces'"}), Collections.emptyMap()).toString());
+        assertEquals("{MESSAGE=argument with spaces}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[] {"ARG", "MESSAGE=\"argument with spaces\""}), Collections.emptyMap()).toString());
+        assertEquals("{TARGETPLATFORM=}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "TARGETPLATFORM"}), Collections.emptyMap()).toString());
+        assertEquals("{TARGETPLATFORM=}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "TARGETPLATFORM="}), Collections.emptyMap()).toString());
+        assertEquals("{TARGETPLATFORM=}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "TARGETPLATFORM:"}), Collections.emptyMap()).toString());
+        assertEquals("{MESSAGE=argument:two}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "MESSAGE=argument:two"}), Collections.emptyMap()).toString());
+        assertEquals("{MESSAGE2=argument=two}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "MESSAGE2=argument=two"}), Collections.emptyMap()).toString());
+        assertEquals("{VER=0.0.3}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "VER=0.0.3"}), Collections.emptyMap()).toString());
+        assertEquals("{VER={0.0.3}}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "VER={0.0.3}"}), Collections.emptyMap()).toString());
+        assertEquals("{VER=[0.0.3]}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "VER=[0.0.3]"}), Collections.emptyMap()).toString());
+        assertEquals("{VER={5,6}}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "VER={5,6}"}), Collections.emptyMap()).toString());
+        assertEquals("{VER={5,6}}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "VER={5,6}"}), Collections.emptyMap()).toString());
+        assertEquals("{VER={}}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "VER={}"}), Collections.emptyMap()).toString());
+        assertEquals("{VER=====}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "VER====="}), Collections.emptyMap()).toString());
+        assertEquals("{MESSAGE=:message}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "MESSAGE=:message"}), Collections.emptyMap()).toString());
+        assertEquals("{MYAPP_IMAGE=myorg/myapp:latest}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "MYAPP_IMAGE=myorg/myapp:latest"}), Collections.emptyMap()).toString());
+        assertEquals("{busyboxVersion=latest}", DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "busyboxVersion"}), Collections.singletonMap("busyboxVersion", "latest")).toString());
     }
 
     @Test(expected = IllegalArgumentException.class)
     public void testInvalidArgWithSpacesFromDockerfile() {
-        DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "MY_IMAGE image with spaces"}));
+        DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "MY_IMAGE image with spaces"}), Collections.emptyMap());
     }
 
     @Test(expected = IllegalArgumentException.class)
     public void testInvalidArgWithTrailingArgumentFromDockerfile() {
-        DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "MESSAGE=foo bar"}));
+        DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "MESSAGE=foo bar"}), Collections.emptyMap());
     }
 
     @Test(expected = IllegalArgumentException.class)
     public void testInvalidArgWithArrayWithSpaceFromDockerfile() {
-        DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "MESSAGE=[5, 6]"}));
+        DockerFileUtil.extractArgsFromLines(Collections.singletonList(new String[]{"ARG", "MESSAGE=[5, 6]"}), Collections.emptyMap());
     }
 
     @Test
     public void testMultiStageNamedWithDuplicates() throws Exception {
         File toTest = copyToTempDir("Dockerfile_multi_stage_named_redundant_build_stages");
         Iterator<String> fromClauses = DockerFileUtil.extractBaseImages(
-                toTest, FixedStringSearchInterpolator.create()).iterator();
+                toTest, FixedStringSearchInterpolator.create(), Collections.emptyMap()).iterator();
 
         assertEquals("centos", fromClauses.next());
         assertFalse(fromClauses.hasNext());
 
+    }
+
+    @Test
+    public void testResolveArgValueFromStrContainingArgKey() {
+        assertEquals("latest", DockerFileUtil.resolveArgValueFromStrContainingArgKey("$VERSION", Collections.singletonMap("VERSION", "latest")));
+        assertEquals("test", DockerFileUtil.resolveArgValueFromStrContainingArgKey("${project.scope}", Collections.singletonMap("project.scope", "test")));
     }
 
     private File copyToTempDir(String resource) throws IOException {

